FreeInit

a-r-r-o-w opened the pull request:
# What does this PR do?

Fixes #6165.

[Colab](https://colab.research.google.com/drive/1NbUkMPf9-2vf51ASZLGAs0cvSNB46FeG?usp=sharing)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [x] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@DN6 @sayakpaul @patrickvonplaten

a-r-r-o-w commented:
This PR attempts to add support for FreeInit. I realize that what I currently have isn't really ideal since it mixes up FreeInit-related code to what should ideally only be AnimateDiff, and limits the maintainability as more features are added in the future. However, I think that adding a new FreeInit-specific pipeline would be unnecessary and lead to a lot of redundant code. Let me know what you think would be the best way to add support for it.

a-r-r-o-w commented:
Sharing some results from current code based on this [notebook](https://colab.research.google.com/drive/1NbUkMPf9-2vf51ASZLGAs0cvSNB46FeG?usp=sharing).

<details>
<summary> Results </summary>

<table>
<tr>
  <td>AnimateDiff</td>
  <td>AnimateDiff + FreeInit (method Gaussian)</td>
  <td>AnimateDiff + FreeInit (method Butterworth)</td>
</tr>
<tr>
  <td colspan=3 align=center> Realistic Vision </td>
</tr>
<tr>
  <td><img src="https://github.com/huggingface/diffusers/assets/72266394/d370665f-891f-4385-aa28-118b1457150f"></td>
  <td><img src="https://github.com/huggingface/diffusers/assets/72266394/7b502f68-c207-49cd-a2ce-21c0eb9ea282"></td>
  <td><img src="https://github.com/huggingface/diffusers/assets/72266394/5a1bd582-5d1b-40db-af40-7c81c2573d6a"></td>
</tr>
<tr>
  <td><img src="https://github.com/huggingface/diffusers/assets/72266394/02b418bc-0b94-4505-91e7-95766df9b806"></td>
  <td><img src="https://github.com/huggingface/diffusers/assets/72266394/326bfe6c-6919-490b-90b8-bd2c856e54bf"></td>
  <td><img src="https://github.com/huggingface/diffusers/assets/72266394/90c3251e-0f40-4d5b-b2c0-a3cee13bd961"></td>
</tr>
<tr>
  <td><img src="https://github.com/huggingface/diffusers/assets/72266394/18783603-9a39-4446-af6e-b4866ac6b67a"></td>
  <td><img src="https://github.com/huggingface/diffusers/assets/72266394/c0f39276-073b-4818-81d9-0700c9cb19e8"></td>
  <td><img src="https://github.com/huggingface/diffusers/assets/72266394/88a6321e-4d26-4352-9b74-64e40326eeca"></td>
</tr>
<tr>
  <td><img src="https://github.com/huggingface/diffusers/assets/72266394/1a5ffd4e-42ff-4948-b0d1-fb55f079cb07"></td>
  <td><img src="https://github.com/huggingface/diffusers/assets/72266394/55c8c563-41d4-4d50-98a8-47c9ebcca01d"></td>
  <td><img src="https://github.com/huggingface/diffusers/assets/72266394/44230be7-aa31-4d20-87df-a83b7881add1"></td>
</tr>
<tr>
  <td colspan=3 align=center> ToonYou </td>
</tr>
<tr>
  <td><img src="https://github.com/huggingface/diffusers/assets/72266394/8f4db2c1-3971-43c1-a349-d43d356ac4d0"></td>
  <td><img src="https://github.com/huggingface/diffusers/assets/72266394/3f16cd93-6dc9-44af-837c-966e2b35001e"></td>
  <td><img src="https://github.com/huggingface/diffusers/assets/72266394/5e241ac5-e892-4374-8823-6819e565e62e"></td>
</tr>
<tr>
  <td><img src="https://github.com/huggingface/diffusers/assets/72266394/5091f39e-d0ac-4637-9aac-94bf06fea52d"></td>
  <td><img src="https://github.com/huggingface/diffusers/assets/72266394/68b2b586-8c1b-4f37-b765-65a24c523cf5"></td>
  <td><img src="https://github.com/huggingface/diffusers/assets/72266394/f475183b-1585-4efe-ac34-ebf440c21249"></td>
</tr>
<tr>
  <td><img src="https://github.com/huggingface/diffusers/assets/72266394/33570d23-0c4a-4980-b141-d69e92f6eede"></td>
  <td><img src="https://github.com/huggingface/diffusers/assets/72266394/b91066e5-47f9-4763-9125-00b24b969907"></td>
  <td><img src="https://github.com/huggingface/diffusers/assets/72266394/15fd398c-e298-44a8-9f64-3405998a1938"></td>
</tr>
</table>
</details>

a-r-r-o-w commented on src/diffusers/pipelines/animatediff/pipeline_animatediff.py:
```diff
@@ -695,52 +727,111 @@ def __call__(
+                device=device,
+                filter_type=freeinit_kwargs.get("method"),
+                n=freeinit_kwargs.get("n", 4) if freeinit_kwargs.get("method") == "butterworth" else None,
+                d_s=freeinit_kwargs.get("d_s", 0.25),
```
Maybe we could rename `d_s` and `d_t` to something like `spatial_stop_frequency` and `temporal_stop_frequency` or better?

a-r-r-o-w commented on src/diffusers/pipelines/animatediff/pipeline_animatediff.py:
```diff
@@ -555,6 +585,8 @@ def __call__(
         cross_attention_kwargs: Optional[Dict[str, Any]] = None,
         clip_skip: Optional[int] = None,
+        use_freeinit: bool = False,
+        freeinit_kwargs: Optional[Dict[str, Any]] = None,
```
```suggestion
        freeinit_kwargs: Optional[Dict[str, Any]] = {},
```
Generation fails otherwise when freeinit is not used due to `.get()` calls to the dict.

DN6 replied:
I would actually try to enable freeinit via a method. Like we have for freeu. e.g. `pipe.enable_freeinit(**free_init_kwargs)`

a-r-r-o-w replied:
Ah okay, that sounds better. Considering the fact that freeinit requires an additional loop that encloses the denoising loop, would you suggest having a separate private method (having freeinit implementation) that `__call__` gets overwritten with? So something like:

```py
def enable_freeinit(self, ...):
    self._original_call = self.__call__
    self.__call__ = self._freeinit_call

def disable_freeinit(self):
    self.__call__ = self._original_call
```

What do you think would be the simplest way to have it added?

DN6 replied:
I think it would be simpler to just move the denoising loop into an internal method and structure it with something like this. We can also add a log message to tell the user that freeinit has been enabled. 

```python
if self.free_init_enabled():
    for fit in free_init_iterations:
          latents = initialize_free_init_args(**relevant free init arguments)
          latents = _denoise(..relevant denoising loop args)
else:
      #default denoising loop 
      latents = _denoise(...)               
```

a-r-r-o-w commented on src/diffusers/pipelines/animatediff/pipeline_animatediff.py:
```diff
@@ -555,6 +585,8 @@ def __call__(
         callback_steps: Optional[int] = 1,
```
IIRC, `callback` and `callback_steps` has been deprecated in favour of `callback_on_step_end` and `callback_on_step_end_tensor_inputs`.

a-r-r-o-w reviewed (commented):
FreeInit changes could be added to AnimateDiff+Controlnet and other future pipelines as well.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_6315). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

a-r-r-o-w commented:
Here's a [Drive](https://drive.google.com/drive/folders/1Ep3epoGn6uXAmSDIpMGuRX81i_ShEVt9?usp=sharing) link to some more results. Also, updated [Colab](https://drive.google.com/file/d/1NbUkMPf9-2vf51ASZLGAs0cvSNB46FeG/view?usp=sharing) notebook.

<details>
<summary>Minimal code</summary>

```py
import torch
from diffusers import MotionAdapter, AnimateDiffPipeline, DDIMScheduler
from diffusers.utils import export_to_gif
from IPython.display import display, Image

adapter = MotionAdapter.from_pretrained("guoyww/animatediff-motion-adapter-v1-5-2")

# model_id = "zhyemmmm/ToonYou"
model_id = "SG161222/Realistic_Vision_V5.1_noVAE"
pipe = AnimateDiffPipeline.from_pretrained(model_id, motion_adapter=adapter, torch_dtype=torch.float16).to("cuda")
pipe.scheduler = DDIMScheduler.from_pretrained(
    model_id,
    subfolder="scheduler",
    beta_schedule="linear",
    clip_sample=False,
    timestep_spacing="linspace",
    steps_offset=1
)
pipe.enable_vae_slicing()

pipe.enable_free_init(method="gaussian", use_fast_sampling=True)
output = pipe(
    prompt=(
        "masterpiece, bestquality, highlydetailed, ultradetailed, sunset, "
        "orange sky, warm lighting, fishing boats, ocean waves seagulls, "
        "rippling water, wharf, silhouette, serene atmosphere, dusk, evening glow, "
        "golden hour, coastal landscape, seaside scenery"
    ),
    negative_prompt="bad quality, worse quality",
    num_frames=16,
    guidance_scale=7.5,
    num_inference_steps=20,
    generator=torch.Generator("cpu").manual_seed(666),
)
pipe.disable_free_init()

frames = output.frames[0]
export_to_gif(frames, "animation.gif")
display(Image("animation.gif", embed=True))
```
</details>

@DN6 I think this is ready for another review :) I've moved the denoising loop and latent decoding part to separate functions in order to allow freeinit to work more easily and also be able to return intermediate results.

DN6 commented on src/diffusers/pipelines/animatediff/pipeline_animatediff.py:
```diff
@@ -43,6 +43,13 @@
 from ..pipeline_utils import DiffusionPipeline
+from .freeinit_utils import (
+    box_low_pass_filter,
+    butterworth_low_pass_filter,
```
Do we need all the additional filters? How do they affect the generation quality? Could we make do with just the default butterworth?

a-r-r-o-w replied:
The mathematical formulation of each of the filters is slightly different, and while butterworth can be used to approximate both ideal and gaussian by setting `order` to a higher or lower value respectively, as documented in the original codebase, it does not seem to produce same/similar results as using the ideal/gaussian filters directly. In the paper, the authors discuss in their ablation study (section C of supplementary material) that while ideal low pass filter provides a clear improvement, it still subjectively fails in regions with large noises. With the gaussian filter, visual quality is improved in comparison to ideal but it can lead to over-smoothed frames. With butterworth, a balance is achieved between visual quality and frame consistency. They also mention the use of gaussian filter for base models with large temporal inconsistency. From testing with low `order` value with butterworth, which should sort of work like gaussian, I wasn't able to get the same effect as using gaussian itself.

The example results I shared above aren't the best results that I've gotten and playing around with `*_stop_frequency` parameters can yield much higher quality/motion. I think that it would be better to stay consistent with the features provided in the original repository/paper, no? Currently, it defaults to sensible parameters but also allows power users to experiment with what works best for a specific model.

The box filter is an approximation of the ideal filter and maybe it could be removed.

a-r-r-o-w replied:
I've refactored this a bit and gotten rid of the freeinit_utils file. Most of the logic for filters are the same and we need to only set mask[..., t, h, w] based on computed mask value. The `def retrieve_mask` part still looks like something that shouldn't make it through review, so do let me know how you'd prefer that. Also dropped `box` filter since it's quality is `inferior` to ideal as it is only an approximation.

DN6 commented on src/diffusers/pipelines/animatediff/pipeline_animatediff.py:
```diff
@@ -539,6 +619,160 @@ def prepare_latents(
+        cross_attention_kwargs,
+        added_cond_kwargs,
+        extra_step_kwargs,
+        callback,
+        callback_steps,
```
We should remove this in favour of `callback_on_step_end`

DN6 commented on src/diffusers/pipelines/animatediff/pipeline_animatediff.py:
```diff
@@ -539,6 +619,160 @@ def prepare_latents(
+
+        return video
+
+    def _retrieve_output(self, latents, output_type, return_dict):
```
Would change this to something more explicit like `_retrieve_video_frames`

DN6 commented on src/diffusers/pipelines/animatediff/pipeline_animatediff.py:
```diff
@@ -539,6 +619,160 @@ def prepare_latents(
+                latents = self._denoise_loop(**denoise_args)
+
+                # Whether or not to return intermediate generation results
+                if self._free_init_return_intermediate_results:
```
Why return intermediate results? Can they be used with a different type of postprocessing?

a-r-r-o-w replied:
So FreeInit is sort of like:

```py
for i in range(freeinit_num_iterations):
    # this gives us back the denoised video sampled at current freeinit iteration
    run_animatediff_completely(noise)
    noise = perform_noise_reinitialization()
```

The intermediate flag allows you to retrieve those intermediate animatediff generations, and perhaps it could be used for comparisons in quality? It has been quite handy to prepare the comparitive table of results in my above comment, and why really throw away a denoised video from the generation process?

DN6 replied:
So would it be fair to say that something like this would give the equivalent results as returning intermediate values?

```python
num_free_init_iterations = 3
videos = []
for free_init_iteration in range(num_free_init_iterations):
      pipe.enable_free_init(num_iters=free_init_iteration)
      video = pipe(**args)
      videos.append(video)
```
My reasoning is that the pipelines should be limited to only what is needed for inference and convenience features should not be included as Pipelines are not intended to be feature-complete user interfaces.

If this is the case and the intermediate values can't be used for anything other than comparison and we can get the equivalent functionality with a for loop, we shouldn't include it inside the Pipeline API.

a-r-r-o-w replied:
That is correct, yes. It does make sense in that way and so I think we can remove it as well now.

a-r-r-o-w replied:
Actually, it would be different when `use_fast_sampling` is set to True since the number of iterations for intermediate results is determined differently for each free_init_iteration. But it still makes sense to remove it since the same behaviour can be obtained by doing things a little differently.

DN6 commented on src/diffusers/pipelines/animatediff/pipeline_animatediff.py:
```diff
@@ -703,55 +938,45 @@ def __call__(
+        }
+
+        if self.free_init_enabled:
+            video = self._free_init_loop(
```
Small nit. Can we just return latents here and pass them to  `_retrieve_video_frames` at the end. Just so the postprocessing isn't hidden inside the loop and both the denoising and free init loops output latents.

a-r-r-o-w replied:
Ah sure.

DN6 reviewed (commented):
Some small changes requested. But overall it's looking good 👍🏽 Can we add a fast test to check that the outputs are different with freeinit enabled.

a-r-r-o-w commented:
@DN6 Requesting a review. I think this should be good to merge, no? I'll open follow up PRs for other animatediff pipelines following this.

a-r-r-o-w commented:
@sayakpaul too, in case some other changes from design pov are required, thanks

DN6 reviewed (commented):
Excellent work @a-r-r-o-w ! I think we just need an update to the AnimateDiff docs (on what FreeInit is and an example snippet) and we can merge this.

DN6 commented on src/diffusers/pipelines/animatediff/pipeline_animatediff.py:
```diff
@@ -79,6 +82,71 @@ def tensor2vid(video: torch.Tensor, processor, output_type="np"):
     return outputs
 
 
+def _get_freeinit_freq_filter(
```
Nice!

DN6 reviewed (approved):
Nice work! 👍🏽

patrickvonplaten commented on src/diffusers/pipelines/animatediff/pipeline_animatediff.py:
```diff
@@ -539,6 +660,185 @@ def prepare_latents(
+
+        return latents
+
+    def _free_init_loop(
```
@DN6 is that ok for you? It adds a bit of double logic to the pipeline and does make it harder to understand - but if ok for you, good for merge for me

a-r-r-o-w replied:
Unfortunately, I do not think there's a way around the extra loop because freeinit requires the additional iteration loop. I've actually refactored out the denoising loop to a separate function to reduce duplication. We can further reduce LoC by refactoring a few things into a separate free_init related file, maybe in a future PR? This might be required because freeinit support can be added across most video generation pipelines and copying ~300-400 lines every time does not seem ideal.

DN6 replied:
I think the intent here was to keep the free init logic separated. FreeInit modifies the noise and latents before running a full denoising loop  every iteration. So when running inference with free init you have an additional progress bar to let you know which freeinit iteration you are currently on.  

I felt having the FreeInit loop wrap the default loop might be confusing since none of that logic is needed when FreeInit is disabled.

patrickvonplaten replied:
Ok for me if you're happy with the design! Feel free to merge whenever!
